File: frontend/src/utils/emailValidation.js
/**
 * Valida se o email pertence a uma lista de provedores confiáveis/permitidos usando um switch case.
 * Provedores permitidos:
 * 1. ufp.edu.pt
 * 2. gmail.com
 * 3. hotmail.com
 * 4. outlook.com
 * 5. outlook.pt
 * 6. yahoo.com
 * 7. yahoo.es
 * 8. yahoo.fr
 * 9. icloud.com
 * 10. sapo.pt
 * 11. live.com.pt
 * 12. live.com
 * 13. aol.com
 * 14. protonmail.com
 * 15. proton.me
 *
 * @param {string} email
 * @returns {{ isValid: boolean, error?: string }}
 */
export function validateEmail(email) {
  if (!email || !email.trim()) {
    return { isValid: false, error: 'Email é obrigatório' };
  }

  const cleanEmail = email.trim().toLowerCase();
  const parts = cleanEmail.split('@');

  if (parts.length !== 2 || !parts[0] || !parts[1]) {
    return { isValid: false, error: 'Formato de email inválido (ex: utilizador@gmail.com)' };
  }

  const domain = parts[1];

  switch (domain) {
    case 'ufp.edu.pt':
    case 'gmail.com':
    case 'hotmail.com':
    case 'outlook.com':
    case 'outlook.pt':
    case 'yahoo.com':
    case 'yahoo.es':
    case 'yahoo.fr':
    case 'icloud.com':
    case 'sapo.pt':
    case 'live.com.pt':
    case 'live.com':
    case 'aol.com':
    case 'protonmail.com':
    case 'proton.me':
      return { isValid: true };
    default:
      return {
        isValid: false,
        error: 'Domínio de email não aceite. Por favor utilize um provedor válido (ex: @ufp.edu.pt, @gmail.com, @outlook.com, @sapo.pt, @hotmail.com, etc.)',
      };
  }
}

File: backend/utils/email_validation.go
package utils

import (
	"net"
	"strings"
)

// LookupMXFunc permite injetar/mockar a resolução de MX em testes unitários.
var LookupMXFunc = net.LookupMX

// IsAllowedDomain verifica através de um switch-case se o domínio pertence à lista de 15 provedores permitidos.
func IsAllowedDomain(domain string) bool {
	domain = strings.ToLower(strings.TrimSpace(domain))
	switch domain {
	case "ufp.edu.pt",
		"gmail.com",
		"hotmail.com",
		"outlook.com",
		"outlook.pt",
		"yahoo.com",
		"yahoo.es",
		"yahoo.fr",
		"icloud.com",
		"sapo.pt",
		"live.com.pt",
		"live.com",
		"aol.com",
		"protonmail.com",
		"proton.me":
		return true
	default:
		return false
	}
}

// HasValidMX confirma se o domínio do email pertence a um provedor permitido e tem registos MX válidos.
func HasValidMX(email string) bool {
	partes := strings.Split(email, "@")
	if len(partes) != 2 || partes[1] == "" {
		return false
	}

	domain := strings.ToLower(strings.TrimSpace(partes[1]))

	// 1. Verificar se é um dos domínios permitidos
	if !IsAllowedDomain(domain) {
		return false
	}

	// 2. Resolver MX
	mxRecords, err := LookupMXFunc(domain)
	if err != nil {
		if dnsErr, ok := err.(*net.DNSError); ok {
			errMsg := strings.ToLower(dnsErr.Error())
			if dnsErr.IsNotFound || strings.Contains(errMsg, "no such host") || strings.Contains(errMsg, "server failure") || strings.Contains(errMsg, "name does not exist") {
				return false
			}
		}
		// Se for erro transitório: fail-open para domínios permitidos
		return true
	}

	return len(mxRecords) > 0
}

// IsDisposableEmail verifica se o email é de um provedor não permitido ou descartável.
func IsDisposableEmail(email string) bool {
	partes := strings.Split(strings.ToLower(strings.TrimSpace(email)), "@")
	if len(partes) != 2 || partes[1] == "" {
		return true
	}

	domain := partes[1]
	return !IsAllowedDomain(domain)
}
